quest_test/test_alias.py

```diff
@@ -0,0 +1,106 @@
+import asyncio
+import pytest
+import logging
+
+from quest import WorkflowManager
+from quest.persistence import InMemoryBlobStorage, PersistentHistory
+from src.quest import step, queue
+from src.quest.historian import Historian
+from src.quest.wrappers import task
+from src.quest.manager_wrappers import alias
+
+# TODO: test exception on alias dict collision
+
+@pytest.mark.asyncio
+async def test_alias():
+    pause = asyncio.Event()
+    data_a = []
+    data_b = []
+
+    async def workflow_a():
+        # TODO: Which format is correct?
+        nonlocal data_a
+        async with alias('the_foo'):
+            async with queue('data', None) as q:
+                # Take alias first
+                data_a.append(await q.get())
+        await pause.wait()
+        data_a.append(await q.get())
+
+    async def workflow_b():
+        nonlocal data_b
+        async with queue('data', None) as q:
+            await pause.wait()
+            # Take alias second
+            async with alias('the_foo'):
+                data_b.append(await q.get())
+            data_b.append(await q.get())
+
+    async def create_workflow(wid: str):
+        match(wid):
+            case 'workflow_a':
+                return workflow_a()
+            case 'workflow_b':
+                return workflow_b()
+
+    storage = InMemoryBlobStorage()
+    histories = {}
+
+    def create_history(wid: str):
+        if wid not in histories:
+            histories[wid] = PersistentHistory(wid, InMemoryBlobStorage())
+        return histories[wid]
+
+    async with WorkflowManager('test_alias', storage, create_history, lambda w_type: workflow_a) as manager:
+        manager.start_workflow('workflow_a', 'wid_a')
+
+        await asyncio.sleep(0.1)
+
+    # async with WorkflowManager('test_alias', storage, create_history, lambda w_type: create_workflow) as manager:
+    #
+    #     # Start both workflows
+    #     manager.start_workflow_background('workflow_a', 'wid_a')
+    #     manager.start_workflow_background('workflow_b', 'wid_b')
+    #
+    #     await asyncio.sleep(0.1)
+    #     # await manager.wait_for_completion('wid_a', None)
+    #     # await manager.wait_for_completion('wid_b', None)
+
+    async with WorkflowManager('test_alias', storage, create_history, lambda w_type: create_workflow) as manager:
+        # Gather resources
+        # TODO: Do I need to grab this again after the alias switches?
+        foo_resources = await manager.get_resources('the_foo', None)
+        a_resources = await manager.get_resources('wid_a', None)
+        b_resources = await manager.get_resources('wid_b', None)
```

Comment:
For this test, I would use the `send_event` function directly:
```py
await manager.send_event('wid_a', 'data', None, 'put', 'data a 1')
await manager.send_event('wid_b', 'data', None, 'put', 'data b 1')
await manager.send_event('the_foo', 'data', None, 'put', 'data foo 1')
```

we may need to change the workflow to something like:
```py
async def workflow_a():
    async with queue('data', None) as q:
        async with alias('the_foo'):
            await first_pause.wait()
            data_a.append(await q.get())
            data_a.append(await q.get())
        await second_pause.wait()
        while q:  # do we have the ability to see if the queue has items?
            data_a.append(await q.get())
```

Do the same for workflow b.

This way we send data to "wid_a", "wid_b", and "the_foo" twice. The first time two items go to workflow A. The second time two items go to workflow B.

At the end, simply assert the full contents of `data_a` and `data_b`:
```py
assert data_a == ['data a 1', 'data foo 1', 'data a 2']
assert data_b == ['data b 1', 'data b 2', 'data foo 2']
```

Reply:
The first pause makes sure the queues are set up before we try to push to them.
The second pause gives the workflows a chance to swap the alias. 

```py
async def workflow_b():
    async with queue('data', None) as q:
            await first_pause.wait()
            data_a.append(await q.get())
        await second_pause.wait()
        async with alias('the_foo'):
            await third_pause.wait()
            data_a.append(await q.get())
            data_a.append(await q.get())
```

Reply:
```py
manager.start('wid_a'...)
manager.start('wid_b'...)
first_pause.set()
await manager.send_event('wid_a', 'data', None, 'put', 'data a 1')
await manager.send_event('wid_b', 'data', None, 'put', 'data b 1')
await manager.send_event('the_foo', 'data', None, 'put', 'data foo 1')
await asyncio.sleep(0.1)  # yield to the workflows

# now both should be waiting on second gate and no one should be the foo
assert not manager.has_workflow('the_foo')
second_pause.set()
await asyncio.sleep(0.1)  # yield

# now workflow b should be the foo
await manager.send_event('wid_a', 'data', None, 'put', 'data a 2')
await manager.send_event('wid_b', 'data', None, 'put', 'data b 2')
await manager.send_event('the_foo', 'data', None, 'put', 'data foo 2')
third_pause.set()
await asyncio.sleep(0.1)  # yield to the workflows

assertions
```